api/layer/layer.go

```diff
@@ -470,72 +471,76 @@ func getRandomOID() (oid.ID, error) {
 	return objID, nil
 }
 
-// DeleteObject removes all objects with the passed nice name.
-func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, obj *VersionedObject) *VersionedObject {
-	if len(obj.VersionID) == 0 {
-		obj.VersionID = UnversionedObjectVersionID
-	}
-	objVersion := &ObjectVersion{
-		BktInfo:               bkt,
-		ObjectName:            obj.Name,
-		VersionID:             obj.VersionID,
-		NoErrorOnDeleteMarker: true,
-	}
-
-	var nodeVersion *data.NodeVersion
-	nodeVersion, obj.Error = n.getNodeVersion(ctx, objVersion)
-
-	if obj.VersionID == UnversionedObjectVersionID {
-		if obj.Error == nil {
-			if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
-				return obj
-			}
-		} else if !errors.IsS3Error(obj.Error, errors.ErrNoSuchKey) {
+func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, settings *data.BucketSettings, obj *VersionedObject) *VersionedObject {
+	if len(obj.VersionID) != 0 || settings.IsNoneStatus {
+		var nodeVersion *data.NodeVersion
+		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
 			return obj
 		}
 
-		randOID, err := getRandomOID()
-		if err != nil {
-			obj.Error = fmt.Errorf("couldn't get random oid: %w", err)
+		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
 			return obj
 		}
 
-		newVersion := &data.NodeVersion{
-			BaseNodeVersion: data.BaseNodeVersion{
-				OID:      randOID,
-				FilePath: obj.Name,
-			},
-			DeleteMarker: &data.DeleteMarkerInfo{
-				Created: time.Now(),
-				Owner:   n.Owner(ctx),
-			},
-			IsUnversioned: true,
-		}
+		obj.Error = n.treeService.RemoveVersion(ctx, bkt.CID, nodeVersion.ID)
+		return obj
+	}
 
-		if obj.Error = n.treeService.AddVersion(ctx, bkt.CID, newVersion); obj.Error != nil {
-			return obj
-		}
+	var newVersion *data.NodeVersion
 
-		n.namesCache.Delete(bkt.Name + "/" + obj.Name)
-	} else {
-		if obj.Error != nil {
-			return obj
-		}
+	if !settings.VersioningEnabled {
```

Comment:
If have neither `null` version nor existing delete marker, the gw doesn't put any delete marker
```
$ aws s3api create-bucket --bucket test8 --endpoint-url http://127.0.0.1:8084
$ aws s3api put-bucket-versioning --versioning-configuration Status=Enabled --bucket test8 --endpoint-url http://127.0.0.1:8084 
$ aws s3api put-object --key a --body ~/Downloads/cat.jpg   --bucket test8 --endpoint-url http://127.0.0.1:8084
$ aws s3api put-object --key a --body ~/Downloads/cat.jpg   --bucket test8 --endpoint-url http://127.0.0.1:8084
$ aws s3api list-object-versions  --bucket test8 --endpoint-url http://127.0.0.1:8084
{
  "Versions": [
    {
      "ETag": "03719ec2b4558dc89c82c47283b4ab9dd17e6e56d1858f254892707491443536",
      "Size": 87422,
      "Key": "a",
      "VersionId": "8tuhfrMiEdvr3m8QkPdE9jVDXgTMTHcVj18RE74rbLtT",
      "IsLatest": true,
      "LastModified": "2022-07-04T16:34:47+00:00",
      "Owner": {
        "DisplayName": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM",
        "ID": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM"
      }
    },
    {
      "ETag": "03719ec2b4558dc89c82c47283b4ab9dd17e6e56d1858f254892707491443536",
      "Size": 87422,
      "Key": "a",
      "VersionId": "F4VwNo7x3PvW9Es29uPu9EMZpnKQ5exTwfGJz34D94HM",
      "IsLatest": false,
      "LastModified": "2022-07-04T16:34:45+00:00",
      "Owner": {
        "DisplayName": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM",
        "ID": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM"
      }
    }
  ]
}
$ aws s3api delete-object --key a --bucket test8  --endpoint-url http://127.0.0.1:8084
$ aws s3api list-object-versions  --bucket test8 --endpoint-url http://127.0.0.1:8084
{
  "Versions": [
    {
      "ETag": "03719ec2b4558dc89c82c47283b4ab9dd17e6e56d1858f254892707491443536",
      "Size": 87422,
      "Key": "a",
      "VersionId": "8tuhfrMiEdvr3m8QkPdE9jVDXgTMTHcVj18RE74rbLtT",
      "IsLatest": true,
      "LastModified": "2022-07-04T16:34:47+00:00",
      "Owner": {
        "DisplayName": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM",
        "ID": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM"
      }
    },
    {
      "ETag": "03719ec2b4558dc89c82c47283b4ab9dd17e6e56d1858f254892707491443536",
      "Size": 87422,
      "Key": "a",
      "VersionId": "F4VwNo7x3PvW9Es29uPu9EMZpnKQ5exTwfGJz34D94HM",
      "IsLatest": false,
      "LastModified": "2022-07-04T16:34:45+00:00",
      "Owner": {
        "DisplayName": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM",
        "ID": "NbUgTSFvPmsRxmGeWpuuGeJUoRoi6PErcM"
      }
    }
  ]
}
```
Logs: 
```
could not delete object {"request_id": "84ca3677-d547-4c89-b5e4-aa91077214ed", "bucket_name": "test8", "object_name": "a", "error": "NoSuchKey: 404 => The specified key does not exist."}
```